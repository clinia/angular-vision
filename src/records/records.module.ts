import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgCviRecords } from './records';

export { NgCviRecords } from './records';

@NgModule({
  declarations: [NgCviRecords],
  entryComponents: [NgCviRecords],
  exports: [NgCviRecords],
  imports: [CommonModule],
})
export class NgCviRecordsModule {}
