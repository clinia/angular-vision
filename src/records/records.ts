import {
  Inject,
  Input,
  Component,
  ContentChild,
  TemplateRef,
  forwardRef,
} from '@angular/core';

import { connectRecordsWithInsights } from '@clinia/vision/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgCviVision, Record } from '../vision/vision';

export interface RecordsState {
  records: Record[];
  results: {};
}

@Component({
  selector: 'cvi-records',
  template: `
    <div [class]="cx()">
      <ng-container *ngTemplateOutlet="template; context: state"></ng-container>

      <!-- default rendering if no template specified -->
      <div *ngIf="!template">
        <ul [class]="cx('list')">
          <li [class]="cx('item')" *ngFor="let record of state.records">
            {{ record.name }}
          </li>
        </ul>
      </div>
    </div>
  `,
})
export class NgCviRecords extends BaseWidget {
  @ContentChild(TemplateRef) public template?: TemplateRef<any>;

  @Input() public escapeHTML?: boolean;
  @Input() public transformItems?: <U extends Record>(items: Record[]) => U[];

  public state: RecordsState = {
    records: [],
    results: {},
  };

  constructor(
    @Inject(forwardRef(() => NgCviVision))
    public visionParent: NgCviVision
  ) {
    super('Records');
  }

  ngOnInit() {
    this.createWidget(connectRecordsWithInsights, {
      escapeHTML: this.escapeHTML,
      transformItems: this.transformItems,
    });
    super.ngOnInit();
  }

  updateState = (state, isFirstRendering: boolean) => {
    if (isFirstRendering) {
      return;
    }
    this.state = state;
  };
}
