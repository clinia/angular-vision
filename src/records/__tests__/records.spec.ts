import { connectRecordsWithInsights } from '@clinia/vision/es/connectors';
import { createRenderer } from '../../../test/test-renderer';
import { NgCviRecords } from '../records';

const defaultState = {
  records: [
    { id: '1', name: 'foo', description: 'foo' },
    { id: '2', name: 'bar', description: 'bar' },
    { id: '3', name: 'foobar', description: 'foobar' },
    { id: '4', name: 'barfoo', description: 'barfoo' },
  ],
  results: { total: 50, page: 0, perPage: 20 },
};

describe('Records', () => {
  it('renders markup without state', () => {
    const render = createRenderer({
      defaultState,
      template: '<cvi-records></cvi-records>',
      TestedWidget: NgCviRecords,
    });
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it('renders markup with state', () => {
    const render = createRenderer({
      defaultState,
      template: '<cvi-records></cvi-records>',
      TestedWidget: NgCviRecords,
    });
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it('should create widget with connectRecords and pass instance options', () => {
    const createWidget = jest.spyOn(NgCviRecords.prototype, 'createWidget');

    const transformItems = jest.fn(x => x);
    const render = createRenderer({
      defaultState,
      template:
        '<cvi-records [escapeHTML]="false" [transformItems]="transformItems"></cvi-records>',
      TestedWidget: NgCviRecords,
      methods: { transformItems },
    });

    render({});

    expect(createWidget).toHaveBeenCalledWith(connectRecordsWithInsights, {
      escapeHTML: false,
      transformItems,
    });
    createWidget.mockRestore();
  });

  it('should expose records to the passed template', () => {
    const render = createRenderer({
      defaultState,
      template: `
          <cvi-records>
            <ng-template let-records="records">
              <ul>
                <li *ngFor="let record of records">
                  {{record.name}}
                </li>
              </ul>
            </ng-template>
          </cvi-records>
        `,
      TestedWidget: NgCviRecords,
    });
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it('should expose results to the passed template', () => {
    const render = createRenderer({
      defaultState,
      template: `
          <cvi-records>
            <ng-template let-results="results">
              {{ results | json }}
            </ng-template>
          </cvi-records>
        `,
      TestedWidget: NgCviRecords,
    });
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });
  it('should allow calling insightsClient', () => {
    const render = createRenderer({
      defaultState,
      template: `
          <cvi-records>
            <ng-template let-records="records" let-insights="insights">
              <ul>
                <li *ngFor="let record of records">
                  <button
                    id="add-to-cart-{{record.id}}"
                    (click)="insights('clickedRecordIDsAfterSearch', { eventName: 'Add to cart', ids: [record.id] })">

                  </button>
                </li>
              </ul>
            </ng-template>
          </cvi-records>
        `,
      TestedWidget: NgCviRecords,
    });

    const insights = jest.fn();
    const fixture = render({ insights });

    const button = fixture.debugElement.nativeElement.querySelector(
      '#add-to-cart-2'
    );
    button.click();

    expect(insights).toHaveBeenCalledWith('clickedRecordIDsAfterSearch', {
      eventName: 'Add to cart',
      ids: ['2'],
    });
  });
});
