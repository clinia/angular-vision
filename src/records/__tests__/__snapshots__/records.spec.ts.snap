// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Records renders markup with state 1`] = `
<ng-component
  testedWidget={[Function NgCviRecords]}
>
  <cvi-vision>
    <cvi-records>
      <div
        class="cvi-Records"
      >
        
        
        <div>
          <ul
            class="cvi-Records-list"
          >
            
            <li
              class="cvi-Records-item"
            >
               foo 
            </li>
            <li
              class="cvi-Records-item"
            >
               bar 
            </li>
            <li
              class="cvi-Records-item"
            >
               foobar 
            </li>
            <li
              class="cvi-Records-item"
            >
               barfoo 
            </li>
          </ul>
        </div>
      </div>
    </cvi-records>
  </cvi-vision>
</ng-component>
`;

exports[`Records renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgCviRecords]}
>
  <cvi-vision>
    <cvi-records>
      <div
        class="cvi-Records"
      >
        
        
        <div>
          <ul
            class="cvi-Records-list"
          >
            
          </ul>
        </div>
      </div>
    </cvi-records>
  </cvi-vision>
</ng-component>
`;

exports[`Records should expose records to the passed template 1`] = `
<ng-component
  testedWidget={[Function NgCviRecords]}
>
  <cvi-vision>
    <cvi-records>
      <div
        class="cvi-Records"
      >
        
        <ul>
          
          <li>
             foo 
          </li>
          <li>
             bar 
          </li>
          <li>
             foobar 
          </li>
          <li>
             barfoo 
          </li>
        </ul>
        
      </div>
    </cvi-records>
  </cvi-vision>
</ng-component>
`;

exports[`Records should expose results to the passed template 1`] = `
<ng-component
  testedWidget={[Function NgCviRecords]}
>
  <cvi-vision>
    <cvi-records>
      <div
        class="cvi-Records"
      >
        
         {
    "total": 50,
    "page": 0,
    "perPage": 20
  } 
        
      </div>
    </cvi-records>
  </cvi-vision>
</ng-component>
`;
